src/devices/Hmc5883l/Hmc5883l.cs

```diff
@@ -0,0 +1,108 @@
+    {
+        /// <summary>
+        /// HMC5883L I2C Address
+        /// </summary>
+        public const byte I2cAddress = 0x1E;
+
+        private I2cDevice _sensor;
+
+        private readonly byte _measuringMode;
+        private readonly byte _outputRate;
+        private readonly byte _gain;
+
+        /// <summary>
+        /// HMC5883L Direction Angle
+        /// </summary>
+        public double DirectionAngle => RawToDirectionAngle(ReadRaw());
+
+        /// <summary>
+        /// Initialize a new HMC5883L device connected through I2C
+        /// </summary>
+        /// <param name="sensor">I2C Device, like UnixI2cDevice or Windows10I2cDevice</param>
+        /// <param name="gain">Gain Setting</param>
+        /// <param name="measuringMode">The mode of measuring</param>
+        /// <param name="outputRate">Typical Data Output Rate (Hz)</param>
+        public Hmc5883l(I2cDevice sensor, Gain gain = Gain.Gain2, MeasuringMode measuringMode = MeasuringMode.Continuous, OutputRate outputRate = OutputRate.Rate5)
+        {
+            _sensor = sensor;
+            _gain = (byte)gain;
+            _measuringMode = (byte)measuringMode;
+            _outputRate = (byte)outputRate;
+
+            Initialize();
+        }
+
+        /// <summary>
+        /// Initialize the sensor
+        /// </summary>
+        private void Initialize()
+        {
+            // Details in Datasheet P12
+            byte configA = (byte)(0x70 + _outputRate << 2);
+            byte configB = (byte)(_gain << 5);
+
+            _sensor.Write(new[] { (byte)Register.HMC_CONFIG_REG_A_ADDR, configA });
+            _sensor.Write(new[] { (byte)Register.HMC_CONFIG_REG_A_ADDR, configB });
+            _sensor.Write(new[] { (byte)Register.HMC_MODE_REG_ADDR, _measuringMode });
+        }
+
+        /// <summary>
+        /// Read raw data from HMC5883L
+        /// </summary>
+        /// <returns>Raw data</returns>
+        public Vector3 ReadRaw()
+        {
+            Span<byte> xRead = stackalloc byte[2];
+            Span<byte> yRead = stackalloc byte[2];
+            Span<byte> zRead = stackalloc byte[2];
+
+            _sensor.Write(new[] { (byte)Register.HMC_X_MSB_REG_ADDR });
+            _sensor.Read(xRead);
+            _sensor.Write(new[] { (byte)Register.HMC_Y_MSB_REG_ADDR });
+            _sensor.Read(yRead);
+            _sensor.Write(new[] { (byte)Register.HMC_Z_MSB_REG_ADDR });
+            _sensor.Read(zRead);
+
+            short x = BinaryPrimitives.ReadInt16BigEndian(xRead);
+            short y = BinaryPrimitives.ReadInt16BigEndian(yRead);
+            short z = BinaryPrimitives.ReadInt16BigEndian(zRead);
+
+            return new Vector3(x, y, z);
+        }
+
+        /// <summary>
+        /// Calculate direction angle
+        /// </summary>
+        /// <param name="rawData">Hmc5883l raw data</param>
+        /// <returns>Angle</returns>
+        public double RawToDirectionAngle(Vector3 rawData)
```

Comment:
On which cases would you be interested in consuming the Raw data? If most of the uses of this would be just to get out the angle, then I would change the surface api to instead just have a ReadDirectionAngle which would perform both operations. That way, the consumer only needs to perform one call, instead of two.

Reply:
Most of the math uses direction vectors and not angles

Reply:
use property **DirectionVector** get raw vector data, **DirectionAngle** get angle, and make this method public